src/librustc/traits/project.rs

```diff
@@ -1320,23 +1320,10 @@ fn assoc_ty_def<'cx, 'gcx, 'tcx>(
     let trait_def_id = selcx.tcx().impl_trait_ref(impl_def_id).unwrap().def_id;
     let trait_def = selcx.tcx().trait_def(trait_def_id);
 
-    if !trait_def.is_complete(selcx.tcx()) {
```

Comment:
I'm having a hard time producing a crash. The following example seems to compile just fine:
```
#![feature(specialization)]

trait Assoc {
    type Output;
}

impl<T> Assoc for T {
    type Output = bool;
}

impl Assoc for u8 {} // <- inherits the non-default type from above
trait Foo {}
impl Foo for u32 {}
impl Foo for <u8 as Assoc>::Output {} // <- this projection will fail

fn main() {
    let _: &Foo = &false;
}
```